### PR TITLE
Refactor the TeamCity config to define all projects, across GA, Beta, and VCR testing

### PR DESCRIPTION
**Pull request by SarahFrench**
# Description

This PR performs a large refactor of the TeamCity configuration for the Google provider. There are artefacts out there describing the project in detail, but here is a summary of the PR's changes:

| Headline | Description | Impact |
|----------|-------------|--------|
| BIG refactor | Lots of files have changed and it's not an easy 1:1 relationship between old and new | Will old files linger in the downstream repo? Not sure |
| A single config for all projects | Previously the TeamCity config described one project, and the config was reused to make multiple projects. Now the config defines all projects needed for testing the GA/Beta Google providers. | We will only use the config in the TPG repo; I've removed .teamcity from the TPGB repo in f1e2e3f. Also, these changes mean that when a new service is added to the provider users will need to manually update .teamcity/components/inputs/services_ga.kt (and _beta.kt) |
| New approach for preventing clashes between builds | In the past we used 'build chains', i.e. explicit dependencies that create pipelines, to force sweepers to run after acceptance tests finish and vice versa. This had some downsides and was confusing. The new approach of using locks to limit the number of builds accessing a service in a GCP project lets acceptance tests run in parallel as long as they test different services, and making sweepers lock all services in a given project helps them run safely. | Users don't need to think about what builds are running or queued when they trigger a build; if builds clash then one will wait in the queue until it's safe. |
| A solution for sweeping google_project resources | Sweeping google_project resources is a pain because they don't respect the GA/Beta/VCR project split that other sweepers benefit from. Using the locks solution above, we can make project sweepers lock ALL services across ALL test projects. I.e. when they run no other builds can run, and vice versa. Also, this sweeper is in it's separate project. | The project sweeper problem is solved and made more prominent in the project. |
| Automated sweeping of VCR test project | This is now automated, whereas in the past users would manually trigger the sweeper | N/A |

# Open questions

- Do we need a VCR recording project that uses the GA provider?

# How do we review?

I think manual testing of some test projects provisioned with the new config will help identify any gaps, see here: https://hashicorp.teamcity.com/project/TerraformProviders_Google_SarahManualTestTeamcityRefactor

If I can get review of the Kotlin code that'd be lovely, but I appreciate there's a lot there. I'm not expecting review like that.

---

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: mmv1/third_party/terraform/.teamcity/components/builds/vcr_build_steps.kt
[the remark this answers is not in the document]
Comment (SarahFrench) [reply mmv1/third_party/terraform/.teamcity/components/builds/vcr_build_steps.kt:1]: Resolved : I made the gcloud and gsutil CLIs available in the AMI used in TeamCity

File: mmv1/provider/terraform/common~compile.yaml
Comment (SarahFrench) [inline mmv1/provider/terraform/common~compile.yaml:1]: Removing this as no TeamCity files are currently being templated or generated using Ruby/ERB

File: GNUmakefile
[the remark this answers is not in the document]
Comment (melinath) [reply GNUmakefile:1]: ugh, that's annoying that it's not possible to generate them easily. That could be a nice future improvement after the go rewrite... but in the meantime doing some kind of validation in the downstream seems reasonable.

File: mmv1/third_party/terraform/.teamcity/tests/build_configuration_features.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
+// This file is controlled by MMv1, any changes made here will be overwritten
```
Comment (melinath) [inline mmv1/third_party/terraform/.teamcity/tests/build_configuration_features.kt:1]: Will we automatically run all these tests somewhere?

File: mmv1/third_party/terraform/.teamcity/PERFORMING_TASKS_IN_TEAMCITY.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
+# How to Use TeamCity
```
Comment (melinath) [inline mmv1/third_party/terraform/.teamcity/PERFORMING_TASKS_IN_TEAMCITY.md:1]: Exciting to have this documented in code! And alongside other TeamCity-related things!

**Review by melinath — APPROVED** [review]
This is a lot of code; I have not validated it in detail but in broad strokes it looks like it has the features we've talked about! I'm also not sure how I would validate it beyond the testing you've already done in TeamCity. Please let me know if you were looking for anything else from me :-)

File: GNUmakefile
[the remark this answers is not in the document]
Comment (SarahFrench) [reply GNUmakefile:1]: Definitely annoying! I've created a separate issue for that validation here (https://github.com/hashicorp/terraform-provider-google/issues/17168), as this PR is already too big

**Comment by SarahFrench** [comment]
> This is a lot of code ... Please let me know if you were looking for anything else from me :-)

@melinath It's definitely too much to read in detail, please don't feel like you need to! I think manual testing in TeamCity is the best option ([here has the config from the PR deployed and set up with creds](https://hashicorp.teamcity.com/project/TerraformProviders_Google_SarahManualTestTeamcityRefactor?projectTab=overview&mode=trends)), or at least locating where you'd perform an action and confirming nothing is missing 😅 

I think the most I'd ask for in terms of looking at the code here is to look at the high level description of what the files in different folder do (mmv1/third_party/terraform/.teamcity/CONTRIBUTION_GUIDE.md) and appreciating that the code is describing _everything_ now, so should be easier for people to connect to what they see in the TeamCity UI. If you think people would be able to do that, given what the code looks like at a glance, I'm happy 😁

File: mmv1/third_party/terraform/.teamcity/CONTRIBUTION_GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+│  │  # There are also files that can supply information about how those packages should be handled, 
+│  │  # e.g. non-default parallelism values that should be used for specific packages.
+│  ├─ projects/
+│  │  # Files containing information about the projects created in the configuration.
+│  │  # Files direcly inside the projects folder define the sub projects that will be created inside
+│  │  # the parent project- Google, Google Beta, and Project Sweeper.
+│  │  # The root_project.kt file defines the parent project where versioned settings is enabled.
+│  │  ├─ reused/
+│  │     # Code for dynamically generating subprojects within Google, Google Beta, and Project Sweeper projects.
+│  ├─ vcsroots/
```
Comment (melinath) [inline mmv1/third_party/terraform/.teamcity/CONTRIBUTION_GUIDE.md:54]: nit:
```suggestion
│  ├─ vcs_roots/
```

**Review by melinath — APPROVED** [review]
yeah, I do believe that a person could figure out what was controlling a particular part of teamcity based on this and make appropriate changes :-)

**Review by shuyama1 — APPROVED** [review]
💯

File: mmv1/third_party/terraform/.teamcity/components/builds/vcr_build_steps.kt [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+                echo "Listing files already present in gs://${'$'}VCR_BUCKET_NAME/beta/fixtures/:"
+                gsutil ls -p ${'$'}GOOGLE_INFRA_PROJECT gs://${'$'}VCR_BUCKET_NAME/beta/fixtures/
+
+                echo "Copying files to Cloud Storage bucket:"
+                gsutil -m cp ${'$'}VCR_PATH/* gs://${'$'}VCR_BUCKET_NAME/beta/fixtures/
+            else
+                echo "Using ${'$'}BRANCH_NAME branch, so copying files to ${'$'}BRANCH_NAME/fixtures/ folder in Cloud Storage bucket"
+
+                gsutil -m cp ${'$'}VCR_PATH/* gs://${'$'}VCR_BUCKET_NAME/${'$'}BRANCH_NAME/fixtures/
```
Comment (shuyama1) [inline mmv1/third_party/terraform/.teamcity/components/builds/vcr_build_steps.kt:178]:
```suggestion
                gsutil -m cp ${'$'}VCR_PATH/* gs://${'$'}VCR_BUCKET_NAME/beta/${'$'}BRANCH_NAME/fixtures/
```

Comment (SarahFrench) [reply mmv1/third_party/terraform/.teamcity/components/builds/vcr_build_steps.kt:178]: Thank you!!

**Comment by SarahFrench** [comment]
I'm just waiting for last nights tests to finish, before getting started on merging this PR and making the changes live

**Comment by SarahFrench** [comment]
The old projects are now deleted and all build history is transferred to the analogous build in the new projects.

**Pull request merged**